src/parcsr_ls/par_strength2nd_device.c

```diff
@@ -0,0 +1,927 @@
+#include "HYPRE.h"
+#include "HYPRE_parcsr_mv.h"
+#include "HYPRE_IJ_mv.h"
+#include "_hypre_utilities.h"
+#include "_hypre_utilities.hpp"
+
+#include "HYPRE_parcsr_ls.h"
+#include "_hypre_parcsr_ls.h"
+#include "_hypre_parcsr_mv.h"
+#include "HYPRE_utilities.h"
+
+#if defined(HYPRE_USING_CUDA)
+
+//-----------------------------------------------------------------------
+HYPRE_Int hypre_BoomerAMGExtractCCDev( hypre_ParCSRMatrix  *A,
+                                       HYPRE_Int           *CF_marker,
+                                       HYPRE_BigInt        *coarse_row_starts,
+                                       hypre_ParCSRMatrix **ACC_ptr);
+
+void truncate( hypre_ParCSRMatrix* S, HYPRE_Complex trunc_level );
+void add_diagonal( hypre_ParCSRMatrix* S );
+void add_unit_data( hypre_ParCSRMatrix* S );
+void remove_diagonal( hypre_ParCSRMatrix* S );
+
+__global__ void hypre_CUDAKernel_ScaleDiag( HYPRE_Int nr_of_rows, HYPRE_Int* SI_diag_i,
+                        HYPRE_Complex* SI_diag_a, HYPRE_Complex coeff );
+
+//-----------------------------------------------------------------------
+HYPRE_Int hypre_BoomerAMGCreate2ndSDevice( hypre_ParCSRMatrix  *S,
+                                           HYPRE_Int           *CF_marker,
+                                           HYPRE_Int            num_paths,
+                                           HYPRE_BigInt        *coarse_row_starts,
+                                           hypre_ParCSRMatrix **C_ptr)
+{
+   HYPRE_Int nr_of_rows hypre_CSRMatrixNumRows(hypre_ParCSRMatrixDiag(S));
+   HYPRE_Complex coeff=1;
+
+   MPI_Comm comm = hypre_ParCSRMatrixComm(S);
+   HYPRE_Int num_proc, myid;
+   hypre_MPI_Comm_size(comm, &num_proc);
+   hypre_MPI_Comm_rank(comm, &myid);
+
+   // 1. Create new matrix with added diagonal, S+coeff*I, for coeff=1 or 2.
+   hypre_NvtxPushRangeColor("Setup", 1);
+   add_unit_data( S );
+   hypre_ParCSRMatrix* SI = hypre_ParCSRMatrixClone( S, 1 );
+   add_diagonal( SI );
+
+   // 2. Add a data array to the strength matrix, needed in order 
+   //    to use matrix-matrix multiplication.
+   add_unit_data( SI );
+   hypre_NvtxPopRange();
+
+   // 2b. Scale diagonal by coeff   
+   if( coeff != 1 )
```

Comment:
@bjorn2 @ulrikeyang Do we need consider `coeff !=1`, which is fixed to be `1` here? My understanding is we need to compute `(S+2I)S` here. So, why `coeff` is 1?

Reply:
My understanding was that it is not completely clear if it should be S+2*I or S+I. However, in the case of num_path=1, I think it makes no difference, so setting coeff=1 saves a few operations. The idea with introducing the variable was that we could easily choose one case or the other.

Reply:
We discussed that S+2*I is needed when setting the number of paths greater than 1. For num_paths == 1 it doesn’t matter what the coefficient is, so in that case S+I is fine. If we can get better performance with S+I, which is equivalent to just the structure, no data values, we should implement that, but it might make sense to also add the option of num_paths > 1 at a later time. Right now, my main concern is to get aggressive coarsening with num_paths 1 working on GPUs.

Reply:
Ulrike, I still don't understand why it is S+2*I, and not S+I, perhaps you could explain ?
Note, the function creat2ndS works also for num_paths==2.
Bjorn

Reply:
It is (S+I)^2-I. Are you saying that using S+I also leads to a correct S2 matrix for num_paths == 2?